add primitive sql support

kevinschoon opened the pull request:
Hi --

Congratulations on writing such an awesome, well-designed, idiomatic Go library!

I made an initial attempt at supporting IO from the standard `database/sql` library; there are definitely some things missing in the code as it exists now but wanted to create a PR and get some feedback. 

Thank you!

tobgu commented:
Thanks! I'll review it in the coming days and come back with feedback!

tobgu commented on internal/io/sql/stmt.go:
```diff
@@ -0,0 +1,29 @@
+	buf.WriteString(table)
+	buf.WriteString("(")
+	for i, name := range colNames {
+		buf.WriteString(name)
```
How would this work with column names that have space in them for example?

tobgu commented on qframe.go:
```diff
@@ -1022,6 +1039,37 @@ func (qf QFrame) ToJSON(writer io.Writer) error {
+	}
+	for i, _ := range qf.index {
+		args := []interface{}{}
+		for j := 0; j < len(columns); j++ {
```
`for _, b := range builders {`?

tobgu commented on qframe.go:
```diff
@@ -1022,6 +1039,37 @@ func (qf QFrame) ToJSON(writer io.Writer) error {
+		return err
+	}
+	for i, _ := range qf.index {
+		args := []interface{}{}
```
A bit of a nitpick:
I've tried to be fairly consistent with initializing slices that have a known size upfront with the size (or capacity) using `make`. Sometimes it has a real performance impact (when working with slices holding column data for example). Sometimes it does not but then I'll still do it for consistency.
In this case I doubt it has any measurable performance impact but it would be nice to do it the same way across the code base.

tobgu commented on internal/io/sql/stmt.go:
```diff
@@ -0,0 +1,29 @@
+		}
+	}
+	buf.WriteString(") VALUES (")
+	for i := 0; i < len(colNames); i++ {
```
`for i := range colNames {`?

tobgu commented on internal/io/sql/reader.go:
```diff
@@ -0,0 +1,96 @@
+// args returns a slice of Scanner for the SQL
+// package. TODO: These pointer allocations seem
+// unnecessary but I'm unsure how else to do it.
+func args(cols []*Column) (ptrs []interface{}) {
```
I've not used named returned types anywhere else in this code base.

I would rather see an explicit allocation of `cols` with a set size by calling `make`.

tobgu commented on qframe.go:
```diff
@@ -1022,6 +1039,37 @@ func (qf QFrame) ToJSON(writer io.Writer) error {
+	columns := qf.ColumnNames()
+	builders := make([]qfsqlio.ArgBuilder, len(columns))
+	for i, name := range columns {
+		builder, err := qfsqlio.NewArgBuilder(qf.columnsByName[name].Column)
```
Nit pick: The assignment could probably be `builders[i], err = ...` to save a line.

tobgu commented on internal/io/sql/reader.go:
```diff
@@ -0,0 +1,96 @@
+// ReadSQL returns a named map of types.DataSlice for consumption
+// by the qframe.New constructor.
+func ReadSQL(rows *sql.Rows) (map[string]types.DataSlice, error) {
+	var columns []*Column
```
What if you, instead of creating a slice of `*Columns` here, created a slice of `interface{}` instead? Then you would never have to do the conversion done in `args(...)` above I guess (when calling `Scan` below)?

kevinschoon commented:
@tobgu Thank you for all the feedback, updates coming ASAP!

kevinschoon commented:
@tobgu I pushed two new commits where I tried to address all of the feedback you left previously, please let me know if I missed anything. 

Major changes include:

* Mocked types for testing SQL transactions (this turned out to be quite a bit more work than expected).
* Improved the column scanning with support for SQL NULL values, various refactoring, and added tests. This is still not really a perfect solution and we might consider using some code generation later on (if only there was some kind of discipline that allowed you to act on _generic_ types.... :slightly_smiling_face:).
* Escaped table name in generated SQL code.
* Fixed various little items mentioned in your review.

tobgu commented:
Cool! I'll try to find some time to review this during the weekend.

tobgu commented on internal/io/sql/reader.go:
```diff
@@ -61,7 +64,7 @@ func (c *Column) Scan(t interface{}) error {
 		c.data.Strings = append(c.data.Strings, string(v))
 	}
-	return nil
+	return errors.New("ReadSQL Scan", fmt.Sprintf("unsupported column type %s", reflect.TypeOf(t).Name()))
```
Picky: No need to call `fmt.Sprintf` here, `errors.New` does the formating for you by doing just that.
This is a general comment.

tobgu commented on internal/io/sql/reader.go:
```diff
@@ -70,13 +73,13 @@ func ReadSQL(rows *sql.Rows) (map[string]types.DataSlice, error) {
 			cols, err := rows.Columns()
 			if err != nil {
-				return nil, err
+				return nil, errors.New("ReadSQL Columns", err.Error())
```
Picky: Consider using `errors.Propagate` here to avoid throwing away the original error.
This is a general comment.

tobgu commented on internal/io/sql/column.go:
```diff
@@ -0,0 +1,133 @@
+	case reflect.Float64:
+		c.data.Floats = append(c.data.Floats, math.NaN())
+	case reflect.String:
+		c.data.Strings = append(c.data.Strings, "")
```
Is this the way you want null-strings to be handled (eg. as empty strings)? Instead of passing `[]string` to `qframe.New()` you could pass `[]*string` to allow nil-values. I'll make sure to add an example of this to the documentation to make it more obvious.

tobgu commented on internal/io/sql/column.go:
```diff
@@ -0,0 +1,133 @@
+		// add any NULL strings previously scanned
+		if c.nulls > 0 {
+			for i := 0; i < c.nulls; i++ {
+				c.data.Strings = append(c.data.Strings, "")
```
See comment about null strings above.

tobgu commented on internal/io/sql/stmt.go:
```diff
@@ -8,17 +8,17 @@ import (
+	buf.WriteString("INSERT INTO \"")
 	buf.WriteString(table)
-	buf.WriteString("(")
+	buf.WriteString("\" (")
```
This seems to address table names with whitespace in them but does it do anything to address column names with whitespace in them?

tobgu commented on qframe_sql_test.go:
```diff
@@ -0,0 +1,159 @@
+	qf := qframe.New(map[string]interface{}{
+		"COL1": []int{1, 2, 3},
+		"COL2": []float64{1.1, 2.2, 3.3},
+		"COL3": []string{"one", "two", "three"},
```
Could you please add a boolean column as well for sake of completeness?

kevinschoon replied:
updated this in my latest changes

tobgu commented on qframe_sql_test.go:
```diff
@@ -0,0 +1,159 @@
+	dvr.results.values = [][]driver.Value{
+		[]driver.Value{int64(1), 1.1, "one"},
+		[]driver.Value{int64(2), 2.2, "two"},
+		[]driver.Value{int64(3), 3.3, "three"},
```
Could you please add a boolean column as well for sake of completeness?

kevinschoon replied:
updated this in my latest changes

kevinschoon commented:
@tobgu Updated the code again based on all your feedback.

Other updates:

* Refactored for more robust database support
* Fixed issues with query generation and added features supporting multiple query types depending on the underlying database.

Have just a few more changes I now need to make to handle some driver-specific peculiarities and then if it all looks good to you it should be ready to merge. 

Here are the three SQL drivers I am aiming to support out of the box:
https://github.com/mattn/go-sqlite3
https://github.com/lib/pq
https://github.com/go-sql-driver/mysql

tobgu commented:
Cool, LGTM! The databases you've chosen should cover most uses that people may have I guess. If not they are free to contribute additional support! :-)

Let me know when you're happy with the branch.

kevinschoon commented:
@tobgu I had to add a little bit more functionality after some testing, this should cover everything now! If you can take one more look, barring any issues you see I am happy with the branch now and it can be merged.

Updates:

* To handle reading and writing booleans from SQLite I added support for coercing values from integers (as they are stored in SQLite) back to bools in QFrame. Pandas has some similar functionality for the same issue. I limited the scope of the changes to only support that specific case but it could also be generalized in the future.
* Added  an option to specify float precision as values are read from a driver. The Postgres driver for instance will accept a float like `1.0` and a more precise value when read back.
* Changed to using prepared queries for reading from the DB.
* Updated the readme example.

Thank you!
